src/USER-HSMA/hsma3d.cpp

```diff
@@ -0,0 +1,1945 @@
+}
+
+void HSMA3D::compute(int eflag, int vflag)
+{
+	// set energy/virial flags
+	ev_init(eflag, vflag);
+	// if atom count has changed, update qsum and qsqsum
+	if (atom->natoms != natoms_original) {
+		qsum_qsq();
+		natoms_original = atom->natoms;
+	}
+	// return if there are no charges
+	if (qsqsum == 0.0) return;
+
+  //Set interfaces
+  double **x = atom->x;
+  double *q = atom->q;
+  int nlocal = atom->nlocal;
+  qqrd2e = force->qqrd2e;
+  double **f = atom->f;
+  double boxlo[3] = { domain->boxlo[0] ,domain->boxlo[1],domain->boxlo[2]};
+  double boxhi[3] = { domain->boxhi[0] ,domain->boxhi[1],domain->boxhi[2]};
+
+  double time;
+  time = MPI_Wtime();
+
+  if (RankID == 1)
+  {
+	  double X[nlocal][3], Q[nlocal], Force[nlocal][3], Pot[nlocal];
+	  for (int i = 0; i < nlocal; i++)
+	  {
+		  X[i][0] = x[i][0] - (boxhi[0] - Lx / 2);
+		  X[i][1] = x[i][1] - (boxhi[1] - Ly / 2);
+		  X[i][2] = x[i][2] - (boxhi[2] - Lz / 2);
+		  Q[i] = q[i];
+		  Force[i][0] = 0.00;
+		  Force[i][1] = 0.00;
+		  Force[i][2] = 0.00;
+		  Pot[i] = 0.00;
+	  }
+	  AdjustParticle_Double(X, nlocal, Lx, Ly, Lz);
+
+	  //Find the image charge
+	  int lx = ceil((Rs - Lx / 2) / Lx), ly = ceil((Rs - Ly / 2) / Ly), lz = ceil((Rs - Lz / 2) / Lz);
+	  int TotalNumber = ceil(nlocal * (2 * lx + 1) * (2 * ly + 1) * (2 * lz + 1) / 2);
+	  double ImageCharge[TotalNumber][4];
+	  int ImageNumber;
+	  SetImageCharge(ImageCharge, &ImageNumber, (2 * lx + 1) * (2 * ly + 1) * (2 * lz + 1), X, Q, nlocal, Rs, Lx, Ly, Lz, lx, ly, lz);
+
+	  //Calculate near field potential which contains all the contributions in the cut-off sphere B at the monitoring points
+	  double Near[Nw];
+	  CalculateNearFieldAndZD_Single(Near, ImageCharge, ImageNumber, Nw, IF_FMM_RightTerm, Lx, Ly, Lz, PointSum, QuizSum, X, Force, Pot, nlocal, Q,tolerance);
+
+	  //Solve The Least Square Problem
+	  double C[p * p];
+	  SolveLeastSquareProblem(C, A, Near, p, Nw);
+
+	  //Compute final force, potential and energy
+	  double Energy_HSMA;
+	  Energy_HSMA = FinalCalculateEnergyAndForce_Single(Force, Pot, X, Q, nlocal, ImageCharge, ImageNumber, Fibonacci, (double**)QRD, (double**)QLocalRD, C, p, Fp, F, Rs, pi, IF_FMM_FinalPotential,tolerance);
+
+	  scale = 1.0;
+	  const double qscale = qqrd2e * scale;
+
+	  //Assemble results
+	  for (int i = 0; i < nlocal; i++)
+	  {
+		  f[i][0] += Force[i][0] * qscale;
+		  f[i][1] += Force[i][1] * qscale;
+		  f[i][2] += Force[i][2] * qscale;
+	  }
+
+	  if (eflag_global) {
+		  energy += Energy_HSMA * qscale / 2;
+	  }
+
+	  if (evflag_atom) {
+		  if (eflag_atom) {
+			  for (int i = 0; i < nlocal; i++) {
+				  eatom[i] += Pot[i];
+				  eatom[i] *= qscale;
+			  }
+		  }
+		  if (vflag_atom)
+			  for (int i = 0; i < nlocal; i++)
+				  for (int j = 0; j < 6; j++) vatom[i][j] *= q[i] * qscale;
+	  }
+  }
+  else if (RankID > 1) {
+
+	  double AllSource[maxatom][3], AllQ[maxatom];
```

Comment:
These kinds of "fortran style" variable length array allocations on the stack are not legal C++. They are accepted as an extension by GCC and a few other compilers, but are rejected when enforcing strict C++. Also, maxatom can become a very large number (LAMMPS supports simulations with more than 2 billion atoms) and such large allocations should not be made on the stack.

In general, arrays of the dimension of the total number of atoms should be avoided at all cost, as they significantly interfere with the efficient strong and weak scaling that LAMMPS achieves using domain decomposition.